### DISCUSSION TITLE
[Question] rationale for PLW1514 autofix / doc suggesting "locale" instead of "utf-8" ?

### DISCUSSION THREAD

**Discussion by mikaelarguedas** [astral-sh/ruff#11364] [Q&A]
Although it's been around for a long time, I discovered PLW1514 with the latest ruff release and have been surprised by the fix which made me curious about it.

ruff docs state the following:
```
Instead, consider using the encoding parameter to enforce a specific encoding. [PEP 597](https://peps.python.org/pep-0597/) recommends using locale.getpreferredencoding(False) as the default encoding on versions earlier than Python 3.10, and encoding="locale" on Python 3.10 and later.
```
And the [autofix implemented](https://github.com/astral-sh/ruff/pull/8928) does the following:
```diff
-open("test.txt")
+open("test.txt", encoding="locale")
```

Reading the [PEP](https://peps.python.org/pep-0597/):
It looks like (from my reading of it):
- the source of the PEP is that without specifying encoding people would assume it is UTF-8 and would get bugs as UTF-8 is not default on Windows
- Python would like to [default to UTF-8](https://peps.python.org/pep-0597/#prepare-to-change-the-default-encoding-to-utf-8) for file opening in the future 
- The `encoding="locale"` is recommended **only when opening text files that have been saved using locale encoding** but the overall recommendation (from my understanding) is to use `encoding="utf-8"`

Finally the example given by ruff differs from the autofix suggested as it gives [utf-8 encoding as an example](https://github.com/qdegraaf/ruff/blob/3b323a09cb5e13d5934c805b860f5c659f967850/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs#L33-L36)

I understand  that `encoding="locale"` is the functional equivalent of not specifying the encoding for now, but from the pep it seems most users would want and potentially should using `utf-8` instead.
So I was wondering if there is a specific rationale for steering users towards using "locale" over "utf-8" ?

Thanks! :pray: 

@Avasam @qdegraaf @charliermarsh

**Comment by Avasam** [comment]
The short answer:
Fixing to "locale" is safer and will keep the code working as-is / as expected. Whilst resolving the lint issue about forward compatibility and implicit encoding.

That being said, I agree that in the documentation, it should steer towards trying to use utf-8 wherever possible.
The autofixed encoding could even be configurable 🤷‍♂️💡


Also after doing this migration in setuptools, I realised that the current suggestion of using `locale.getpreferredencoding(False)` on older python doesn't help at all resolving EncodingWarning as you just end up with a different warning. Imo it should instead be `"locale" if sys.version >=(3,10) else None`
